### ISSUE TITLE
Update these Typings with DefinitelyTyped

### ISSUE THREAD

**Issue by simon-xiong-ca** [louy/typed-mongoose#8]
Hey, I maintain the mongoose definitions for DefinitelyTyped https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/mongoose

Could we update these typings to match the one on DefinitelyTyped?
I have a few more changes I need to make to them. I'll send a PR when I'm done.

**Comment by simon-xiong-ca** [comment]
There is a fix in it for #4

**Comment by wayne-werner** [comment]
DT's mongoose.d.ts is completely incompatible with code written to this d.ts.  I still can't figure out how I'm supposed to type the result of a connection.model<t> or extend it so that Schema statics can be recognized as part of the new model by intellisensing.  This is quite simple and direct with this d.ts.

I plan to keep using this d.ts.  Maybe dt can take a few pointers and update from here as well?

Both of you guys are doing a great deal of hard work.  Thank you.  A merging of your efforts would be great, but honestly, I would not be happy if this d.ts became a clone of the dt's d.ts.  As things stand now, I would lose the ease being able to use the Model interface defined here, as well as other issues I have with the dt d.ts like the ease of replacing promise with the global promise found here as well.

**Comment by simon-xiong-ca** [comment]
@wayne-oscme We're not going to drag and drop it in here. Do you have specific examples of code that breaks with DT's mongoose.d.ts so we can make it better?

**Comment by simon-xiong-ca** [comment]
@wayne-oscme Well actually maybe we were lol. Anyways I'm holding off until I have some free time. If you would like to use the one on DefinitelyTyped, but have problems with it you can open up an issue so we can improve it.

**Comment by louy** [comment]
Hey! I've done so in #10. There was only one issue:

I liked the hack for `MongoosePromise` (#4) but I think it's a little bit trickier to do in this case. I'll have to give it a try before actually using it. If anyone can show a working demo PRs are always welcome.

**Comment by simon-xiong-ca** [comment]
@louy Oh wow nice job! I guess we're going with this then. There's some more updates pending for the DT mongoose.d.ts https://github.com/DefinitelyTyped/DefinitelyTyped/pull/10455 I'll take a look at #10 in a bit and figure out if Promises can work with the typings structure. I'll also help with the issues that come rolling in when we update.

**Comment by simon-xiong-ca** [comment]
@wayne-oscme `connection.model(name, schema)` returns `mongoose.Model`. If you want to add properties/methods to docs returned by queries, you can do: 

```
interface IDoc {
  prop: string;
  method: () => void;
}
var Model: mongoose.Model<IDoc> = connection.model<IDoc>(...);
Model.findOne({}, (err, doc) => {
  doc.prop;         // works
  doc.method();     // works
};
```

If you want statics: 

```
interface IStatics {
  staticProp: string;
  staticMethod: () => void;
}
var Model: IStatics & mongoose.Model<IDoc> = connection.model<IDoc, IStatics>(...);
Model.staticProp;         // works
Model.staticMethod();     // works
```

**Comment by wayne-werner** [comment]
Here's how I was doing it using this d.ts:

```
interface MyDocument extends Document {
  //...
  prop: string;
  method: () => void;
}
interface MyModel extends Model<MyDocument> {
  //...
  staticProp: string;
  staticMethod: () => void;
}

var mySchema = new Schema({
  //...
});

// Make a model from the document interface and the schema.
export var Final:MyModel = <MyModel>connection.model<MyDocument>('Final', mySchema); 

```

A tad more obvious than what you're doing, I couldn't figure out how to extend model and quiet the typescript compiler about unknown members of my model for the statics using the dt version...  I'm going to give it a try and see how it goes.  Example code should be a part of any type repository, I think.

BTW, the dt version has a few unmentioned dependencies, can't remember them all.  mongdb.d.ts being one.

Let me see what's what, and I'll get back on my resolution.  Thanks for the leg up, and thanks to both groups for some seriously hard work!

**Comment by wayne-werner** [comment]
FYI, about those missing dependencies:

```
typings INFO reference Stripped reference "https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/d4724f8c3e9e11335770556fef3e5ee7265e36c1/mongodb/mongodb.d.ts" during installation from "mongoose" (main)
typings INFO reference Stripped reference "https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/d4724f8c3e9e11335770556fef3e5ee7265e36c1/mpromise/mpromise.d.ts" during installation from "mongoose" (main)
typings INFO reference Stripped reference "https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/d4724f8c3e9e11335770556fef3e5ee7265e36c1/node/node.d.ts" during installation from "mongoose" (main)
typings INFO reference Stripped reference "https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/d4724f8c3e9e11335770556fef3e5ee7265e36c1/mongoose-promise/mongoose-promise.d.ts" during installation from "mongoose" (main)
```

**Comment by wayne-werner** [comment]
Sweet... I've got it all going with the dt version...  I think anyone using the current npm version, though, are  going to have a couple of hiccups like I did.

I found this really useful, both for variable typing and downcasting the result of a .model call:

```
import {Model} from 'mongoose';
//...
export type MyStuffModel = MyStuffStatics & Model<MyStuffDocument>; 
export const MyStuff: MyStuffModel  = <MyStuffModel>connection.model<MyStuffDocument>('MyStuff', schema);
```

I like to export the model, as opposed to using .model('string') as it can be obtained, without the need to downcast.  If you're already importing the type for variable typing, might as well just fetch model at the same time.

Thanks again.

Question:  Going forward, after the two projects merge, should I have typings point to the npm or dt registry for this d.ts?

**Comment by louy** [comment]
I'm having a few problems switching from this implementation to the one in DT and I think they're because the latter uses union types.

I find the solution we did in [typed-sequelize](https://github.com/louy/typed-sequelize#usage) to be much nicer than using union types. I think the only challenge with mongoose is `_id` as it's a string in `interface Thing` and an instance of `ObjectID` in `interface ThingInstance` and there's no way to override attribute types without unions.

What do you guys think? I can't merge #10 until all of the issues are fixed or otherwise it would be a breaking change. I also think switching to interfaces instead of union types won't be a breaking change and that's why I'm suggesting it here.

**Comment by wayne-werner** [comment]
It was that issue that got me

Another breaking change:
Missing signature in `Model<doc>` for property `modelName`.

**Comment by simon-xiong-ca** [comment]
@wayne-oscme Just curious, how come you don't export your Model like this:

```
import {Model} from 'mongoose';
//...
export const MyStuff: MyStuffStatics & Model<MyStuffDocument> = connection.model<MyStuffDocument, MyStuffStatics>('MyStuff', schema);
```

From what I know DT is organized as global modules so I would imagine you'd have to jump through some hoops if you're switching from npm or vice versa.

**Comment by simon-xiong-ca** [comment]
@louy By union types do you mean 1) `type1 & type2` or 2) `type1 | type2`? In my mind I'm thinking 1). Correct me if I'm wrong. 

Also, how do people test these typings definitions? For DT, we have a test file like this: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/mongoose/mongoose-tests.ts where we write a bunch of code that should compile without errors using the definitions.

@wayne-oscme Mongoose in particular has a very flexible public interface (Ex: you can swap out the promise library, extend Schemas with your own methods, statics, and 3rd party plugins). It's hard to represent this in definitions without using some of our own arbitrary conventions.

If we can work out the quirks, I can put together a doc on what those conventions are and how to use them.

**Comment by wayne-werner** [comment]
Okay.  First, the type definition is just a way to make that ugly union a part of the model implementation, not the interface of the model.  Just to make it easier to use outside of the model definition file.  But the problem is real.

What if I'm in a different file, say myrouter.ts.

```
import {Connection, Document} 'mongoose';
// get the connection and such, then...
var myModel = connection.model<Document>('MyStuff');
```

Typescript can't make the inferences.  `myModel.myStatic` is an error now.  None of my statics appear in the ide code completion.

I'm really not going to make the user of the model remember a composite type for the model, and to have import the interfaces. No way.  So, 

```
export type MyStuffModel = MyStuffStatics & Model<MyStuffDocument>; 
export const MyStuff: MyStuffModel  = <MyStuffModel>connection.model<MyStuffDocument>('MyStuff', schema);

```

and

```
import {Connection, Document} 'mongoose';
import (MyStuffModel) from './my-stuff';
// get the connection and such, then...
var myModel = <MyStuffModel>(connection.model<Document>('MyStuff'));
```

Best, if I'm not processing models in a loop, and I know what I want:

```
import (MyStuff) from './my-stuff';
// just use MyStuff.
```

I'm not exactly sure what your objection to the type def is.

What if I want to have the type in an interface?  A class member? Wherever I need the type.  One of the nice things about the npm dt is that the types are just interfaces extending Model and Document.  No friction.  No cryptic type unions in type errors either!

**Comment by simon-xiong-ca** [comment]
@louy @wayne-oscme That's a good point. I can bring interfaces back. I have one suggestion:
This relates to how we implement the definitions and will not affect users. Mongoose uses 3 interfaces `Model`, `model`, and `Document`.

Their relationships are:
- `model` extends `Document` and `EventEmitter`
- `Model` extends `EventEmitter`
- `new Model()`, `Model.findOne()`, etc. returns a `model` not a `Document`, more specifically `Document` does not have a `.save()` method http://mongoosejs.com/docs/api.html#document-js but we put it in there to patch it up, which I don't agree with.

I think we should have 2 separate interfaces for `model` and `Document` otherwise, it doesn't follow Mongoose's code structure and is harder to maintain.

But, since people have been using this forever:

```
interface IDoc extends Document {
  ...
}
```

We should probably keep using `Document` as the type returned from `Model.find()` for backwards compatibility. Could we keep those interfaces separate and use: `interface Document extends model { ... }` so we know which method belongs to which interface?

**Comment by simon-xiong-ca** [comment]
I'll be updating DT to try to make the switch from union types to interfaces. I would suggest putting  #10 on hold for now.

**Comment by louy** [comment]
That's what I was trying to do yesterday and I got really confused. I really don't like they way mongoose does some stuff, and this is one of them.

One of the issues I had is with `SubDocument` and `Embedded`. They have a totally different signature for `#remove` which means that we can't `extend Document` for these two.

I'll leave #10 for now as you suggested. Let me know once you've done these changes (even if they aren't merged yet) and I can take a look at them/merge them into this one.

**Comment by simon-xiong-ca** [comment]
@louy I've made a patch here https://github.com/simonxca/DefinitelyTyped/blob/patch-mongoose/mongoose/mongoose.d.ts

Changes:
- Got rid of the top level mongoose namespaces that wrap everything. Tests still work, don't think it was necessary (sorry, more work for you if you copied them over!)
- I changed from using type references and `type1 & type2` to just using the interfaces from the old mongoose definitions. @wayne-oscme In particular I made sure code like this works:

```
interface MyDocument extends mongoose.Document {
  prop: string;
  method: () => void;
}
interface MyModel extends mongoose.Model<MyDocument> {
  staticProp: string;
  staticMethod: () => void;
}
interface ModelStruct {
  doc: MyDocument;
  model: MyModel;
  method1: (callback: (model: MyModel, doc: MyDocument) => void) => MyModel;
}
var modelStruct1: ModelStruct;
var myModel1: MyModel;
var myDocument1: MyDocument;
modelStruct1.method1(function (myModel1, myDocument1) {
  myModel1.staticProp;
  myModel1.staticMethod();
  myDocument1.prop;
  myDocument1.method();
}).staticProp;
```
- For Document, I followed Mongoose's structure of `mongoose.model extends mongoose.Document`. This will separate `mongoose.model` methods from `mongoose.Document` methods. For backwards compatibility, I did a rename so it's now `mongoose.Document extends mongoose.MongooseDocument`. Not sure about this, what's your opinion?

Let me know if there's anything you see that needs changing.

**Comment by louy** [comment]
That's great! thank you.

Two things:
- the `import mongoose = require('mongoose');` trick doesn't work with `typings`. It's a circular reference and it breaks the generated file.
- `mpromise` has no method `catch` or `addErrback` which are used in the test file for some reason.

**Comment by louy** [comment]
A few more issues caused by this:
- Return type of remove queries shouldn't be `void` but [`WriteOpResult`](https://docs.mongodb.com/manual/reference/method/db.collection.remove/#definition). (simonxca/DefinitelyTyped#2)
- `Model.collection` and `Model.modelName` are missing. (simonxca/DefinitelyTyped#1)
- signature for `Model#create` is messed up a little bit. (not sure if we can do something about this but it breaks a lot of stuff in my projects)

**Comment by wayne-werner** [comment]
The only time I need to do `import mongoose = require('mongoose');` is to set `mongoose.Promise = global.Promise;`.  Otherwise, `import {<whatever>} from 'mongoose';` works just fine for me to get access to whatever I need.   Can't wait till I don't need to do any require.  I'd kind of thought getting away from `mongoose.whatever` is considered a good thing.

I'm running into the same `Model.modelName` and `Model.collection` issues.

@louy could you give a link to Model#create in the documentation?  I can't seem to find it.  I'd really like to take look at the defined signature vs documented signature.

@simonxca Wow.  Fast work.  Thanks!

**Comment by simon-xiong-ca** [comment]
@louy Thanks, I'll take a look over the weekend, have to finish some stuff this week.

**Comment by simon-xiong-ca** [comment]
@louy Hmm, you're right I'm not sure how to handle `mongoose.Mongoose` yet, will have to take a look at that. Do you have an alternative?
I believe `catch` and `addErrback` are from Mongoose's implementation (which inherit from mpromise) here: http://mongoosejs.com/docs/api.html#promise_Promise-catch

**Comment by simon-xiong-ca** [comment]
@louy Could you post some code snippets of how you use `Model.create`?
@wayne-oscme Documentation is here: http://mongoosejs.com/docs/api.html#model_Model.create

It creates and saves documents for you. I assumed when you called `Model.create()`, whatever you plugged into `mongoose.model<...>` would be the type returned in the callback/promise. The signature in the definitions might be a bit off. I couldn't figure out how to add types into the callback.

**Comment by wayne-werner** [comment]
Sorry for being unresponsive... real life obligations. 

I've been using the version at "https://raw.githubusercontent.com/simonxca/DefinitelyTyped/patch-mongoose/mongoose/mongoose.d.ts".  Should I continue with this version?

Only hitch I've encountered in the past week is that `Document` is missing `__v?: number` declaration.

**Comment by simon-xiong-ca** [comment]
@wayne-oscme yup it should be good. You're right, I'll add that in.
